### DISCUSSION TITLE
CircuitPython v  MicroPython - any plans to add USB drive support as in CircuitPython

### DISCUSSION THREAD

**Discussion by FlydGitHub** [micropython/micropython#11066] [RP2040 / Pico]
I am a great fan of MicroPython, I think it's totally awesome.

IMHO it lacks one feather of CircuitPython, and that when the uf2 file is used,  CircuitPython comes back as a USB drive, whereas in MicroPython, after the uf2 file is written, it's only available in a editor, like Thonny or Vscode.

**Comment by robert-hh** [comment]
This topic is discussed a lot in the community. MSC support exists at the moment for Pyboard and by default for Arduino Nano 2040 connect. There at Pull request for supporting it for a few other boards. So the software components exist. But as convenient as it seems, it has issues. First of all, for support the board has to use a FAT file system. That's not a good file system with flash chips. Then, unless the firmware uses an exclusive access mechanism like CircuitPython there is a high risk of file system corruption sooner or later. 
There are file access methods like MTP, which is used e.g. by smartphones, which do not have these restrictions. But there seems no "micro" implementation available.
Saying that, as a user I expect no soon decision about that topic. Until then, tools like mpremote, rshell in general or ftp for boards with WiFi/Ethernet work fine and reliable.

**Comment by peterhinch** [comment]
See [this note](https://github.com/peterhinch/micropython-samples/blob/master/README.md#15-a-usb-pitfall) which explains why MSC mode can lead to file corruption.

**Comment by FlydGitHub** [comment]
Thank you Robert and Peter, your diligence is very much appreciated.

**Reply by robert-hh** [reply]
One last thing to tell; mpremote offers a different approach for file exchange with the PC. It allows to mount a directory of the PC at the board. The files on the PC appear as local files of the board, under the path name /remote, AFAIK. That works well and has not risk of file corruption, since each side has consistent control.

**Reply by peterhinch** [reply]
Indeed. I can't recommend it highly enough. Its transfer speed is astonishing!
